src/computability/NFA.lean

```diff
@@ -0,0 +1,131 @@
+/-
+Copyright (c) 2020 Fox Thomson. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Fox Thomson
+-/
+
+import data.fintype.basic
+import computability.DFA
+
+/-!
+# Nondeterministic Finite Automata
+This file contains the definition of a Nondeterministic Finite Automaton (NFA), a state machine
+which determines whether a string (implemented as a list over an arbitrary alphabet) is in a regular
+set by evaluating the string over every possible path.
+We show that DFA's are equivalent to NFA's however the construction from NFA to DFA uses an
+exponential number of states.
+-/
+
+universes u v
+
+/-- An NFA is a set of states (`state`), a transition function from state to state labelled by the
+  alphabet (`step`), a starting state (`start`) and a set of acceptance states (`accept`).
+  Note the transition function sends a state to a `finset` of states. These are the states that it
+  may be sent to. -/
+structure NFA (alphabet : Type u) (state : Type v) :=
+(step : state → alphabet → finset state)
+(start : finset state)
+(accept : finset state)
+
+namespace NFA
+
+variables {α : Type u} {σ σ' σ₁ σ₂ σ₃ : Type v} (M : NFA α σ)
+variables [decidable_eq σ] [decidable_eq σ₁] [decidable_eq σ₂] [decidable_eq σ₃]
+
+instance NFA_inhabited : inhabited (NFA α σ') := ⟨ NFA.mk (λ _ _, ∅) ∅ ∅ ⟩
+
+/-- `M.step_set S a` is the union of `M.step s a` for all `s ∈ S`. -/
+def step_set : finset σ → α → finset σ :=
+λ Ss a, finset.bind Ss (λ S, (M.step S a))
+
+lemma mem_step_set (s : σ) (S : finset σ) (a : α) :
+  s ∈ M.step_set S a ↔ ∃ t ∈ S, s ∈ M.step t a :=
+by rw [step_set, finset.mem_bind]
+
+/-- `M.eval_from S x` computes all possible paths though `M` with input `x` starting at an element
+  of `S`. -/
+def eval_from (start : finset σ) : list α → finset σ :=
+list.foldl M.step_set start
+
+/-- `M.eval x` computes all possible paths though `M` with input `x` starting at an element of
+  `M.start`. -/
+def eval := M.eval_from M.start
+
+/-- `M.accepts x` says that there is an accept state in `M.eval x`. -/
+def accepts (s : list α) : Prop :=
+∃ S ∈ M.accept, S ∈ M.eval s
+
+/-- Two NFA's are equivalent if they accept exactly the same strings. -/
+def equiv (M : NFA α σ₁) (N : NFA α σ₂) : Prop := ∀ x, M.accepts x ↔ N.accepts x
```

Comment:
Note that by `propext` and `funext`, you could also declare this as
```suggestion
def equiv (M : NFA α σ₁) (N : NFA α σ₂) : Prop := M.accepts = N.accepts
```
which may or may not be helpful.

Reply:
This seems to be making reflexivity, symmetry and transitivity harder to prove, `equiv_refl` is no longer `by refl` but `by tauto`.